### DISCUSSION TITLE
Next13 Middleware Stuck on Login Page After Successful Login with Custom signIn Page

### DISCUSSION THREAD

**Discussion by lko3001** [nextauthjs/next-auth#9716] [Help]
### Question 💬

I'm using Next13 and Auth.js, I keep getting redirected to the login page even though the login is successful. I want it to redirect me to the page that I was planning to go to, for example, when I try to go to the `/todo` page, I get redirected to the login page `http://localhost:3000/signin?callbackUrl=%2Ftodo` but when I try to login I get redirected to `http://localhost:3000/signin?callbackUrl=%2Ftodo`, the same route. Why? It only happens when I use a custom `signIn` page.

I've already looked a lot of posts about this error, I tried all of the solutions but none of them worked, for example I tried giving the provider an `id` but it only caused more problems, the only thing that kind of worked was adding this to the NextAuth options:
```ts
 callbacks: {
    async signIn({ user, account, profile, email, credentials }) {
      return true;
    },
    async redirect({ url, baseUrl }) {
      console.log({ url, baseUrl });
      return baseUrl;
    },
  },
```
But it always redirected me to the homepage obviously, but I want it to redirect me to the page I was directed to, but I can't return the `url` from that function because the `console.log` logs this, so it would just redirect me to the signIn page and I can't figure out how to fix it.
```js
{ url: 'http://localhost:3000', baseUrl: 'http://localhost:3000' }
{
  url: 'http://localhost:3000/signin?callbackUrl=%2Ftodo',
  baseUrl: 'http://localhost:3000'
}
{ url: 'http://localhost:3000', baseUrl: 'http://localhost:3000' }
```

Just for context here's some code:

This is my `src\app\api\auth\[...nextauth]\route.ts`
```ts
import NextAuth from "next-auth";
import { authOptions } from "./options";

const handler = NextAuth(authOptions);
export { handler as GET, handler as POST };
```
Which imports `src\app\api\auth\[...nextauth]\options.ts` file:
```ts
import { NextAuthOptions } from "next-auth";
import GitHubProvider from "next-auth/providers/github";
import { PrismaAdapter } from "@auth/prisma-adapter";
import { prisma } from "@/prisma/client";
import { Adapter } from "next-auth/adapters";
import NextAuth, { Session } from "next-auth";

export const authOptions: NextAuthOptions = {
  secret: process.env.NEXTAUTH_SECRET,
  adapter: PrismaAdapter(prisma) as Adapter,
  providers: [
    GitHubProvider({
      clientId: process.env.GITHUB_ID!,
      clientSecret: process.env.GITHUB_SECRET!,
    }),
  ],
  session: {
    strategy: "jwt",
  },
  jwt: {
    secret: process.env.NEXTAUTH_SECRET,
  },
  pages: {
    signIn: "/signin",
  },
};
```
And then my `src/middleware.ts` is like this:
```ts
import { withAuth } from "next-auth/middleware";

export default withAuth({
  pages: {
    signIn: "/signin",
  },
});

export const config = { matcher: "/todo" };
```

### How to reproduce ☕️

Since I don't know what causes the bug I'm not even sure how to fix it since the working version should just redirect me to the desired page, so I can't help you reproduce the issue but I'm willing to post make the repo public.

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by ybhwang** [comment]
Same problem :(

**Comment by njoguamos** [comment]
Same problem when using QwikJs.

**Comment by ALVNCHNG** [comment]
I'm also experiencing the same issue when using Auth0 as one of the providers. (I've also used Prisma as an Adapter)

**Comment by goerlitz** [comment]
I faced the same issue. It happens when you define a custom login page in `src\app\api\auth\[...nextauth]\options.ts`.
```TypeScript
  pages: {
    signIn: "/signin",
  },
```

I simply put following code at the beginning of my custom login page (a server component):
```TypeScript
import { getServerSession } from 'next-auth';
import { redirect } from 'next/navigation';
import { authOptions } from '@/app/api/auth/[...nextauth]/options';

export default async function SignIn() {
  const session = await getServerSession(authOptions);
  if (session) {
    redirect('/');
  }
```

This gets you away from login page - but without using the callbackUrl parameter.
~~I'm curious how it is done in the built-in login page, but I have not figured it out yet.~~

UPDATE: below is a working solution.

**Comment by goerlitz** [comment]
Eventually, I found the solution.
The relevant part of the documentation is here: https://next-auth.js.org/getting-started/client#specifying-a-callbackurl.

Basically, you need to take the callback url from the current url search parameters and give it as a parameter to the SignIn function.

This needs to be done in a client component. I my case, the login page is a server component and only the social login button is a client component.
```typescript
'use client';

import { signIn } from 'next-auth/react';
import { useSearchParams } from 'next/navigation';
import Button from '@mui/material/Button';
import GoogleIcon from '@mui/icons-material/Google';

export default function SocialLogin() {
  const searchParams = useSearchParams();

  return (
    <Button
      fullWidth
      variant="outlined"
      onClick={() => {
        signIn('google', { callbackUrl: searchParams.get('callbackUrl') || undefined });
      }}
    >
      <GoogleIcon fontSize="small" />
      Sign in with Google
    </Button>
  );
}
```

Hope this works for other people, too.

**Comment by jennnx** [comment]
I am getting the same error in Next.js14 + NextAuth v5 as well (5.0.0-beta.3).

As author mentioned, only when I'm using: Custom signIn page _and_ Next.js middleware together.

EDIT: I got the solution fixed by following https://github.com/nextauthjs/next-auth/issues/8978#issuecomment-1794890212

1. Stop your server
2. Delete the .next folder
3. Manually add AUTH_URL=http://localhost:3000/api/auth (or your domain, when you're in production)

And it resolve it for me.

The auth object in the middleware now shows up correctly

**Comment by vlasaAndrei** [comment]
I have the same issue and I cannot find a solution, neither. It redirects all routes back to the signin page.

**Comment by goerlitz** [comment]
@BIGVl , you need to provide more details. Otherwise, nobody can help you.

How does your `singin` function call look like? It has to include the desired callback Url, like this example.

`signIn('google', { callbackUrl: searchParams.get('callbackUrl') || '/dashboard' });`
